src/VisualStudio/CSharp/Test/ProjectSystemShim/CSharpReferencesTests.cs

```diff
@@ -0,0 +1,119 @@
+                project2.OnImportAdded(@"c:\project1.dll", "project1");
+
+                Assert.Equal(true, project2.GetCurrentProjectReferences().Any(pr => pr.ProjectId == project1.Id));
+
+                project2.Disconnect();
+                project1.Disconnect();
+            }
+        }
+
+        [Fact]
+        [Trait(Traits.Feature, Traits.Features.ProjectSystemShims)]
+        public void AddCylicProjectMetadataReferences()
+        {
+            using (var environment = new TestEnvironment())
+            {
+                var project1 = CreateCSharpProject(environment, "project1");
+                environment.ProjectTracker.UpdateProjectBinPath(project1, null, @"c:\project1.dll");
+
+                var project2 = CreateCSharpProject(environment, "project2");
+                environment.ProjectTracker.UpdateProjectBinPath(project2, null, @"c:\project2.dll");
+
+                project1.AddProjectReference(project2);
+
+                // normally this metadata reference would be elevated to a project reference, but fails because of cyclicness
+                project2.OnImportAdded(@"c:\project1.dll", "project1");
+
+                Assert.Equal(true, project1.GetCurrentProjectReferences().Any(pr => pr.ProjectId == project2.Id));
+                Assert.Equal(false, project2.GetCurrentProjectReferences().Any(pr => pr.ProjectId == project1.Id));
+
+                project2.Disconnect();
+                project1.Disconnect();
+            }
+        }
+
+        [Fact]
+        [Trait(Traits.Feature, Traits.Features.ProjectSystemShims)]
+        public void AddCylicProjectReferences()
+        {
+            using (var environment = new TestEnvironment())
+            {
+                var project1 = CreateCSharpProject(environment, "project1");
+                var project2 = CreateCSharpProject(environment, "project2");
+
+                project1.AddProjectReference(project2);
+                project2.AddProjectReference(project1);
+
+                Assert.Equal(true, project1.GetCurrentProjectReferences().Any(pr => pr.ProjectId == project2.Id));
+                Assert.Equal(false, project2.GetCurrentProjectReferences().Any(pr => pr.ProjectId == project1.Id));
+
+                project2.Disconnect();
+                project1.Disconnect();
+            }
+        }
+
+        [Fact]
+        [Trait(Traits.Feature, Traits.Features.ProjectSystemShims)]
+        public void AddCylicProjectReferencesDeep()
+        {
+            using (var environment = new TestEnvironment())
+            {
+                var project1 = CreateCSharpProject(environment, "project1");
+                var project2 = CreateCSharpProject(environment, "project2");
+                var project3 = CreateCSharpProject(environment, "project3");
+                var project4 = CreateCSharpProject(environment, "project4");
```

Comment:
That's what the AddCylicProjectMetadataReferences test is testing, since cross language references appear as metadata references that we try to convert to project references. To the ProjectTracker that maintains the set of all projects, there is no difference between languages, there just needs to be a known project with the correct ID for the test to work.